daemon/cmd/endpoint.go

```diff
@@ -731,13 +733,62 @@ func (d *Daemon) EndpointDeleted(ep *endpoint.Endpoint, conf endpoint.DeleteConf
 	}
 }
 
+func (d *Daemon) setIngressIPsFromEndpoint(ep *endpoint.Endpoint) {
+	changed := false
+	// Use EPs addresses as ingress addresses
+	if ingressIPv4 := node.GetIngressIPv4(); ingressIPv4 == nil && ep.IPv4.IsSet() {
+		ip := ep.IPv4.IP()
+		if !ip.Equal(node.GetIngressIPv4()) {
+			node.SetIngressIPv4(ip)
+			changed = true
+		}
+	}
+	if ingressIPv6 := node.GetIngressIPv6(); ingressIPv6 == nil && ep.IPv6.IsSet() {
+		ip := ep.IPv6.IP()
+		if !ip.Equal(node.GetIngressIPv6()) {
+			node.SetIngressIPv6(ip)
+			changed = true
+		}
+	}
+	// Update any possible Envoy listeners that may depend on the changed values
+	if changed && d.l7Proxy != nil {
+		ctx, cancel := context.WithTimeout(context.Background(), option.Config.EnvoyConfigTimeout)
+		wg := completion.NewWaitGroup(ctx)
+		err, revertFuncs := d.l7Proxy.UpdateListenersForNodeChanges(wg)
+		if err != nil {
+			log.WithError(err).Error("Envoy Listener updates for ingress address changes failed")
+			return
+		}
+		// Wait & revert asynchronously
+		go func() {
+			defer cancel()
+			if len(revertFuncs) > 0 {
+				start := time.Now()
+				log.Debug("Waiting for proxy updates to complete...")
+				err := wg.Wait()
+				log.Debugf("Wait time for proxy updates %v", time.Since(start))
+
+				// revert all changes in case of failure
+				if err != nil {
+					revertFuncs.Revert(nil)
+					log.WithError(err).Warn("daemon: Finished reverting failed Envoy Listener updates for ingress address changes")
+				}
+				return
+			}
+		}()
+	}
+}
+
 // EndpointCreated is a callback to satisfy EndpointManager.Subscriber,
 // allowing the EndpointManager to be the primary implementer of the core
 // endpoint management functionality while deferring other responsibilities
 // to the daemon.
 //
 // It is called after Daemon calls into d.endpointManager.AddEndpoint().
 func (d *Daemon) EndpointCreated(ep *endpoint.Endpoint) {
+	if ep.IsIngressDrone() {
```

Comment:
We didn't yet broach the subject of "how do we plumb the networking for this endpoint?". Right now it looks like we're plumbing it through to the network, potentially routing traffic towards it, then just assuming that if we plumb the ingress bits correctly then no traffic will arrive to that endpoint.

If any assumption there is wrong, then I think you'll end up getting RSTs and ICMP Port Unreachables coming back from that endpoint. I wonder whether that makes more sense, or whether we could put a BPF program in that just drops traffic with a dedicated error code "BUG: Ingress traffic was misrouted". This would at least be a bit easier to notice in drop metrics for instance (particularly during restarts / upgrades).